subsys/net/lib/download_client/src/coap.c

```diff
@@ -44,6 +49,42 @@ int coap_block_init(struct download_client *client, size_t from)
 	return 0;
 }
 
+int coap_get_snd_timeout(struct download_client *dl)
+{
+	if (has_pending(dl)) {
+		return dl->coap.pending.timeout;
+	}
+
+	return CONFIG_COAP_INIT_ACK_TIMEOUT_MS;
```

Comment:
I wonder, shouldn't this be treated as an error condition? I see that the current implmentation calls `coap_pending_init()` and `coap_pending_cycle()`, as expected, so it shouldn't be the case that we try to recv/send w/o  a proper timeout? If we return some default value here, such an even might undergo unnoticed.

Reply:
If recv was okay, there is no pending object and therefore we send coap request using CONFIG_COAP_INIT_ACK_TIMEOUT_MS. The time send blocks is substracted again in coap_get_recv_timeout, so that we wait per cycle in total the timeout given by coap implementation. To make it short, the code is fine :)

Reply:
> there is no pending object and therefore we send coap request using CONFIG_COAP_INIT_ACK_TIMEOUT_MS.

So the question is why there is no **new** pending object for the next request, if we already attempt to send it.

I've tried to understand the current logic, and it seems that the callstack looks like this:
```
send_block_request() - here we configure the send timeout, based on uninitialized `coap.pending`
         |
         v     
   request_send()
         |
         v    
 coap_request_send() - and only here we initialize `coap.pending`
         |
         v    
   socket_send() - we could set the timeout here as well, just before send()
         |
         v    
       send() - called in a loop
```

I don't understand why do we try to use the `coap.pending` structure before it's properly initialized. Because of it we need workarounds like this. IMO it'd be cleaner if the `coap.pending` structure was the only source for the timeout value (especially that using `CONFIG_COAP_INIT_ACK_TIMEOUT_MS` may not be accurate if we randomize the ACK timeout for instance).

Reply:
Thanks a lot. I totally agree. I have changed the point a new pending object is initialized. I simply took the code from poll-implementation but in this case a coap request is initialized on every coap_request_send and so a coap pending object already exists. Using poll-mechanism required a completely different approach